go/vt/vttablet/onlineddl/executor.go

```diff
@@ -3384,6 +3426,58 @@ func (e *Executor) executeMigration(ctx context.Context, onlineDDL *schema.Onlin
 	return nil
 }
 
+// getNonConflictingMigration finds a single 'ready' migration which does not conflict with running migrations.
+// Conflicts are:
+// - a migration is 'ready' but is not set to run _concurrently_, and there's a running migration that is also non-concurrent
+// - a migration is 'ready' but there's another migration 'running' on the exact same table
+func (e *Executor) getNonConflictingMigration(ctx context.Context) (*schema.OnlineDDL, error) {
+	pendingMigrationsUUIDs, err := e.readPendingMigrationsUUIDs(ctx)
+	if err != nil {
+		return nil, err
+	}
+	r, err := e.execQuery(ctx, sqlSelectReadyMigrations)
+	if err != nil {
+		return nil, err
+	}
+	for _, row := range r.Named().Rows {
+		uuid := row["migration_uuid"].ToString()
+		onlineDDL, migrationRow, err := e.readMigration(ctx, uuid)
+		if err != nil {
+			return nil, err
+		}
+		isImmediateOperation := migrationRow.AsBool("is_immediate_operation", false)
+
+		if conflictFound, _ := e.isAnyConflictingMigrationRunning(onlineDDL); conflictFound {
+			continue // this migration conflicts with a running one
+		}
+		if e.countOwnedRunningMigrations() >= maxConcurrentOnlineDDLs {
+			return nil, nil // too many running migrations
+		}
+		if isImmediateOperation && onlineDDL.StrategySetting().IsInOrderCompletion() {
+			// This migration is immediate: if we run it now, it will complete within a second or two at most.
+			if len(pendingMigrationsUUIDs) > 0 && pendingMigrationsUUIDs[0] != onlineDDL.UUID {
+				continue
+			}
+		}
+		// We will fail an in-order migration if there's _prior_ migrations within the same migration-context
+		// which have failed.
+		if onlineDDL.StrategySetting().IsInOrderCompletion() {
+			wasFailed, err := e.validateInOrderMigration(ctx, onlineDDL)
+			if err != nil {
+				return nil, err
+			}
+			if wasFailed {
+				continue
+			}
+		}
```

Comment:
Most of this function is just an extract/refactor out of `runNextMigration()` and into an independent function of its own. But these lines above are then also added to fail a migration before even running it, if the in-order terms are not met.